lib/SILOptimizer/Mandatory/TFPartition.cpp

```diff
@@ -1548,6 +1548,10 @@ static bool hoistValueAboveStartPoint(SILInstruction *inst,
   if (DI.properlyDominates(inst, tensorStartPoint))
     return true;
 
+  // It doesn't make sense to hoist the start point above itself.
```

Comment:
an alternative fix is to call `DI.properlyDominates()` above with `DI.Dominates()` -- that'd be more consistent with the impl of the dual function `sinkValueAfterEndPoint()`, which also calls `Dominates()`.

WDYT?

Reply:
Do you mean
```
  if (DI.dominates(inst, tensorStartPoint))
    return true;
```
?

That won't work, because it's important to return `false`. For example, in the example in the PR description, `true` will make it move `builtin "cmp_eq_Int32"` above `integer_literal $Builtin.Int32, 1`.

The inconsistency comes from the fact that `tensorStartPoint` is "the first Tensor operation" and `tensorEndPoint` is "the first instruction after the last Tensor operation". If we made the definitions more consistent (e.g. make `tensorStartPoint` be "the last instruction before the first Tensor operation"), then these functions could be more consistent. But that would probably be a big overhaul.

Reply:
You are right. Sorry about the confusion.
The patch looks good!